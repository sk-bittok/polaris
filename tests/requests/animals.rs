use crate::{request, requests::prepare_auth};

use axum::http::StatusCode;
use insta::{Settings, assert_debug_snapshot, with_settings};
use polaris::{models::animals::Animal, views::animals::AnimalResponse};
use serial_test::serial;

macro_rules! configure_insta {
    ($(expr:expr),*) => {
        let mut settings = Settings::clone_current();
        settings.set_snapshot_path("snapshots/animals");
        settings.set_prepend_module_to_snapshot(false);
        settings.set_snapshot_suffix("anim");
        let _guard = settings.bind_to_scope();
    };
}

#[tokio::test]
#[serial]
async fn can_get_all() {
    request(|server, ctx| async move {
        configure_insta!();

        crate::seed_data(&ctx.db).await.unwrap();

        let user = prepare_auth::login_user(&server, &ctx).await;

        let (auth_header, auth_value) = prepare_auth::auth_header(user.access_token);

        let request = server
            .get("/animals")
            .add_header(auth_header, auth_value)
            .await;

        assert_eq!(request.status_code(), StatusCode::OK);

        let data = request.json::<Vec<AnimalResponse>>();

        assert_debug_snapshot!((request.status_code(), data));
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_get_by_id() {
    request(|server, context| async move {
        configure_insta!();

        crate::seed_data(&context.db).await.unwrap();

        let user = prepare_auth::login_user(&server, &context).await;
        let (auth_header, auth_value) = prepare_auth::auth_header(user.access_token);

        let id = "f6417c11-d817-4626-9e8d-c68a44002d4b";
        let request = server
            .get(&format!("/animals/{id}"))
            .add_header(auth_header, auth_value)
            .await;

        assert_debug_snapshot!((request.status_code(), request.text()));
    })
    .await
}

#[tokio::test]
#[serial]
async fn can_post_one() {
    request(|server, context| async move {
        configure_insta!();

        crate::seed_data(&context.db).await.unwrap();

        let user = prepare_auth::login_user(&server, &context).await;
        let (auth_header, auth_value) = prepare_auth::auth_header(user.access_token);

        let params = serde_json::json!({
           "tagId": "AC025",
           "name": "Geoffrey",
           "specie": "cattle",
           "breed": "Friesian",
           "gender": "male",
           "status": "active",
           "currentWeight": 25000,
           "purchasePrice": 3500000,
           "purchaseDate": "2025-02-18",
           "dateOfBirth": "2023-10-16"
        });

        let request = server
            .post("/animals")
            .json(&params)
            .add_header(auth_header, auth_value)
            .await;

        assert_eq!(request.status_code(), StatusCode::CREATED);

        with_settings!({
            filters => {
                let mut filters = crate::cleanup_date().to_vec();
                filters.extend(crate::cleanup_uuid().to_vec());
                filters.extend(crate::cleanup_int().to_vec());
                filters
            }
        }, {
            assert_debug_snapshot!((request.status_code(), request.json::<Animal>()));
        })
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_delete_one() {
    request(|server, context| async move {
        configure_insta!();

        crate::seed_data(&context.db).await.unwrap();
        let user = prepare_auth::login_user(&server, &context).await;
        let (auth_header, auth_value) = prepare_auth::auth_header(user.access_token);

        let id = "f6417c11-d817-4626-9e8d-c68a44002d4b";

        let request = server
            .delete(&format!("/animals/{}", id))
            .add_header(auth_header, auth_value)
            .await;

        assert_debug_snapshot!((request.status_code(), request.text()));
    })
    .await;
}

#[tokio::test]
#[serial]
async fn can_get_by_tag_id() {
    request(|server, context| async move {
        configure_insta!();

        crate::seed_data(&context.db).await.unwrap();

        let user = prepare_auth::login_user(&server, &context).await;
        let (auth_header, auth_value) = prepare_auth::auth_header(user.access_token);

        let id = "AC003";

        let request = server
            .get(&format!("/animals/tag-id/{}", id))
            .add_header(auth_header, auth_value)
            .await;

        assert_debug_snapshot!((request.status_code(), request.text()));
    })
    .await;
}
