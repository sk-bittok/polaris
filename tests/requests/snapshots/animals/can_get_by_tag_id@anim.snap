---
source: tests/requests/animals.rs
assertion_line: 152
expression: "(request.status_code(), request.text())"
---
(
    200,
    "{\"id\":103,\"pid\":\"d909e761-36da-4062-ae78-abba4f7c1103\",\"organisationPid\":\"9d5b0c1e-6a48-4bce-b818-dc8c015fd8a0\",\"organisationName\":\"Acme Corp\",\"tagId\":\"AC003\",\"name\":\"Ferdinand\",\"specieName\":\"cattle\",\"breedName\":\"Jersey\",\"dateOfBirth\":\"2022-08-20\",\"gender\":\"male\",\"parentFemaleName\":null,\"parentFemaleTagId\":null,\"parentFemaleId\":null,\"parentMaleName\":null,\"parentMaleTagId\":null,\"parentMaleId\":null,\"status\":\"active\",\"purchaseDate\":\"2022-10-05\",\"purchasePrice\":\"65000.00\",\"weightAtBirth\":\"32.25\",\"currentWeight\":\"625.75\",\"notes\":\"Breeding bull, excellent lineage\",\"createdBy\":\"bd6f7c26-d2c9-487e-b837-8f77be468033\",\"createdByName\":\"John Doe\",\"createdAt\":\"2024-10-06T08:45:00Z\",\"updatedAt\":\"2024-12-10T11:30:00Z\"}",
)
